### PR TITLE
[refactor] OSS tests

### PR DESCRIPTION
**Pull request by blefaudeux**
# Before submitting
- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Improves on #311, for instance the world size and the devices are a bit cleaner I think. The tests use pytorch infrastructure on purpose, to ease upstreaming, since I found that to be a pain point

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by blefaudeux** [comment]
Based on the elastic checkpoints branch to minimize future conflicts, but it means that not all changes here are related to this PR. Only changing the unit tests

**Comment by blefaudeux** [comment]
Not ready, hangs on CI.. this is infinite

File: tests/optim/test_oss.py [hunk 1]
```diff
@@ -117,672 +137,549 @@ def test_lr_scheduler(self):
...
-
+            # Update the optimizer state on the reference rank
+            optimizer.consolidate_state_dict(recipient_rank=reference_rank)
+
+            # Fetch the state on the reference rank
+            # - check that it has the correct size
+            # - load it again
+            if self.rank == reference_rank:
+                optimizer_state_dict = optimizer.state_dict()
+                # self.assertEqual(len(optimizer_state_dict["state"]), len(list(model.parameters())))
```
Comment (blefaudeux) [inline tests/optim/test_oss.py:357]: this depends on #310, will need to be uncommented out

**Comment by blefaudeux** [comment]
except for the test which is adapted to #310, the rest of the tests are unchanged, same as before

**Comment by anj-s** [comment]
@blefaudeux Looks like this PR is in draft mode. Do you still want me to review it?

**Comment by blefaudeux** [comment]
> @blefaudeux Looks like this PR is in draft mode. Do you still want me to review it?

sorry about that, the CI is still not happy so there are things I should fix before review :/ looks like the test util that I'm using changed in between torch 1.5/1.6/17, maybe that I should use fairscale.testing instead

**Comment by blefaudeux** [comment]
maybe reviving this at some point, I cannot reproduce this one locally and I feel like it's not worth it really..

**Pull request closed without merging**